### PR TITLE
Fix align of intrinsic return.

### PR DESCRIPTION
**Pull request by EnderGnar**
Fixed issue that the alignment was chosen wrongly when an intrinsic was called with a return place.

### CODE REVIEW

File: spec/lang/step.md [hunk 1]
```diff
@@ -554,16 +554,21 @@ impl<M: Memory> Machine<M> {
...
         let arguments = arguments.try_map(|arg| self.eval_value(arg))?;
 
-        let (value, ret_type) = self.eval_intrinsic(intrinsic, arguments)?;
+        let (value, ret_ty) = self.eval_intrinsic(intrinsic, arguments)?;
 
         if let Some(ret_place) = ret_place {
-            let align = Align::max_for_offset(ret_type.size::<M>()).unwrap();
-            
-            let bytes = Type::encode::<M>(ret_type, value);
+            let place_ty = place_ty.unwrap();
```
Comment (RalfJung) [inline spec/lang/step.md:567]: I think you can avoid the unwrap if you do the ret_expr.check_wf inside the `if let` here.

File: spec/lang/step.md [hunk 2]
```diff
@@ -554,16 +554,21 @@ impl<M: Memory> Machine<M> {
...
+
+            let align = place_ty.align;
+
+            let bytes = Type::encode::<M>(ret_ty, value);
             self.mem.store(ret_place, bytes, align)?;
```
Comment (RalfJung) [inline spec/lang/step.md:572]: This should use `typed_store`

File: spec/lang/step.md (hunk 1 above)
Comment (EnderGnar) [reply spec/lang/step.md:567]: Hmm but ret_expr would still need to be unwrapped. the only thing I would see is to just do `if let Some= ret_expr`. but then `eval_place` happens after evaluating the arguments.

Comment (RalfJung) [reply spec/lang/step.md:567]: Sure, but this `map` outside the `if` should not be needed.

File: spec/lang/step.md [hunk 3]
```diff
@@ -558,13 +558,13 @@ impl<M: Memory> Machine<M> {
...
 
-        let (value, ret_type) = self.eval_intrinsic(intrinsic, arguments)?;
+        let (value, _ret_ty) = self.eval_intrinsic(intrinsic, arguments)?;
 
         if let Some(ret_place) = ret_place {
-            let align = Align::max_for_offset(ret_type.size::<M>()).unwrap();
-            
-            let bytes = Type::encode::<M>(ret_type, value);
-            self.mem.store(ret_place, bytes, align)?;
+            let ret_expr = ret_expr.unwrap();
```
Comment (RalfJung) [inline spec/lang/step.md:564]: Oh now I see what you mean... we should really just make `eval_place` return the type. But that's for a future change.

File: spec/lang/step.md [hunk 4]
```diff
@@ -558,13 +558,13 @@ impl<M: Memory> Machine<M> {
         // Evaluate all arguments.
         let arguments = arguments.try_map(|arg| self.eval_value(arg))?;
 
-        let (value, ret_type) = self.eval_intrinsic(intrinsic, arguments)?;
+        let (value, _ret_ty) = self.eval_intrinsic(intrinsic, arguments)?;
```
Comment (RalfJung) [inline spec/lang/step.md:561]: I didn't realize this means we entirely ignore the `ret_ty` here. That's not good. Argh why are function calls so annoying.^^ I'll have to think about what we want to do here.

We should definitely add a test for calling an intrinsic where the result is put into a place of the wrong type.

File: spec/lang/step.md (hunk 3 above)
Comment (EnderGnar) [reply spec/lang/step.md:564]: Wouldn't this introduce a lot of redundancy?

File: spec/lang/step.md (hunk 4 above)
Comment (EnderGnar) [reply spec/lang/step.md:561]: Isn't it the case that we also don't check anything about types when a function returns? I have just written a test where a u64 is returned into a u8 and it passes.

File: spec/lang/step.md (hunk 3 above)
Comment (RalfJung) [reply spec/lang/step.md:564]: Yes. But that might be the lesser evil.

File: spec/lang/step.md (hunk 4 above)
Comment (RalfJung) [reply spec/lang/step.md:561]: That sounds potentially concerning, could you open an issue and paste the example there?

**Pull request merged**